File: Data_Structure/Linked_list/CircularLL.java
import java.util.Scanner;

public class CircularListDemo {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Linklist list = new Linklist();

        while (true) {
            System.out.println("\nMenu:");
            System.out.println("1. Insert at the Start");
            System.out.println("2. Insert at the End");
            System.out.println("3. Display List");
            System.out.println("4. Get List Size");
            System.out.println("5. Exit");
            System.out.print("Enter your choice: ");

            int choice = scanner.nextInt();
            switch (choice) {
                case 1:
                    System.out.print("Enter value to insert at the start: ");
                    int value1 = scanner.nextInt();
                    list.insertAtFirst(value1);
                    break;
                case 2:
                    System.out.print("Enter value to insert at the end: ");
                    int value2 = scanner.nextInt();
                    list.insertAtLast(value2);
                    break;
                case 3:
                    System.out.println("Current List: ");
                    list.displayList();
                    break;
                case 4:
                    System.out.println("List Size: " + list.getSize());
                    break;
                case 5:
                    System.out.println("Exiting...");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice. Please enter a number between 1 and 5.");
            }
        }
    }
}

class Linklist {
    class Node {
        int data;
        Node next;

        public Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node last;
    private int size;

    public Linklist() {
        this.last = null;
        this.size = 0;
    }

    public boolean isEmpty() {
        return last == null;
    }

    public int getSize() {
        return size;
    }

    public void insertAtFirst(int val) {
        Node n = new Node(val);
        if (isEmpty()) {
            last = n;
            last.next = last;
        } else {
            n.next = last.next;
            last.next = n;
        }
        size++;
    }

    public void insertAtLast(int val) {
        Node n = new Node(val);
        if (isEmpty()) {
            last = n;
            last.next = last;
        } else {
            n.next = last.next;
            last.next = n;
            last = n;
        }
        size++;
    }

    public void displayList() {
        if (isEmpty()) {
            System.out.println("The list is empty.");
            return;
        }
        Node temp = last.next;
        do {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        } while (temp != last.next);
        System.out.println("(back to start)");
    }
}

File: Data_Structure/Linked_list/CircularLL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Linklist:
    def __init__(self):
        self.last = None
        self.size = 0

    def is_empty(self):
        return self.last is None

    def get_size(self):
        return self.size

    def insert_at_first(self, val):
        new_node = Node(val)
        if self.is_empty():
            self.last = new_node
            self.last.next = self.last
        else:
            new_node.next = self.last.next
            self.last.next = new_node
        self.size += 1

    def insert_at_last(self, val):
        new_node = Node(val)
        if self.is_empty():
            self.last = new_node
            self.last.next = self.last
        else:
            new_node.next = self.last.next
            self.last.next = new_node
            self.last = new_node
        self.size += 1

    def display_list(self):
        if self.is_empty():
            print("The list is empty.")
            return
        temp = self.last.next
        while True:
            print(temp.data, end=" -> ")
            temp = temp.next
            if temp == self.last.next:
                break
        print("(back to start)")

if __name__ == "__main__":
    import sys

    list = Linklist()

    while True:
        print("\nMenu:")
        print("1. Insert at the Start")
        print("2. Insert at the End")
        print("3. Display List")
        print("4. Get List Size")
        print("5. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            value1 = int(input("Enter value to insert at the start: "))
            list.insert_at_first(value1)
        elif choice == 2:
            value2 = int(input("Enter value to insert at the end: "))
            list.insert_at_last(value2)
        elif choice == 3:
            print("Current List: ")
            list.display_list()
        elif choice == 4:
            print("List Size: ", list.get_size())
        elif choice == 5:
            print("Exiting")
            sys.exit()
        else:
            print("Invalid choice. Please enter a number between 1 and 5.")
